javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -0,0 +1,455 @@
+        return containsKey(currentElement._1) ? (M) remove(currentElement._1).put(newElement) : (M) this;
+    }
+
+    @Override
+    public M replaceAll(Tuple2<K, V> currentElement, Tuple2<K, V> newElement) {
+        return replace(currentElement, newElement);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M retainAll(Iterable<? extends Tuple2<K, V>> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        return (M) createFromEntries(back.flatMap(t -> t._2.map(v -> Tuple.of(t._1, v))).retainAll(elements));
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M scan(Tuple2<K, V> zero, BiFunction<? super Tuple2<K, V>, ? super Tuple2<K, V>, ? extends Tuple2<K, V>> operation) {
+        Objects.requireNonNull(operation, "operation is null");
+        return (M) Collections.scanLeft(this, zero, operation, Queue.<Tuple2<K, V>>empty(), Queue::append, this::createFromEntries);
+    }
+
+    @Override
+    public Iterator<Multimap<K, V>> sliding(long size) {
+        return sliding(size, 1);
+    }
+
+    @Override
+    public Iterator<Multimap<K, V>> sliding(long size, long step) {
+        return iterator().sliding(size, step).map(this::createFromEntries);
+    }
+
+    @Override
+    public Tuple2<Multimap<K, V>, Multimap<K, V>> span(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final Tuple2<Iterator<Tuple2<K, V>>, Iterator<Tuple2<K, V>>> t = iterator().span(predicate);
+        return Tuple.of(createFromEntries(t._1), createFromEntries(t._2));
+    }
+
+    @Override
+    public M tail() {
+        if (isEmpty()) {
+            throw new UnsupportedOperationException("tail of empty Multimap");
+        } else {
+            final Tuple2<K, V> head = head();
+            return remove(head._1, head._2);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M take(long n) {
+        if (size() <= n) {
+            return (M) this;
+        } else {
+            return (M) createFromEntries(iterator().take(n));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M takeRight(long n) {
+        if (size() <= n) {
+            return (M) this;
+        } else {
+            return (M) createFromEntries(iterator().takeRight(n));
+        }
+    }
+
+    @Override
+    public M takeUntil(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return takeWhile(predicate.negate());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M takeWhile(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final Multimap<K, V> taken = createFromEntries(iterator().takeWhile(predicate));
+        return taken.length() == length() ? (M) this : (M) taken;
+    }
+
```

Comment:
XOR is bad function for hashCode. For example, HAMT of 1000 integers all between 0 and 16 will also between 0 and 16.